### PR TITLE
Add a script method to get its class icon

### PR DESCRIPTION
**Pull request by YuriSizov**
This is more of a proof-of-concept built on top of https://github.com/godotengine/godot/pull/75472, but perhaps we can polish it into something mergeable? The goal here is to extract the icon for built-in scripts, so we can use it in the editor like we do with regular script classes.

This exposes an editor-only method to fetch an icon path of the script's class. The reason why I'm still skeptical here, is because this doesn't allow for multiple icons per one script (in case it contains multiple classes). Perhaps the API should be generalized somehow, maybe by passing a class name to this method.

But maybe it's fine as is, and we only accept one icon per file. This is totally acceptable for GDScript anyway, at least for now.

image

### CODE REVIEW

File: doc/classes/ScriptExtension.xml [hunk 1]
```diff
@@ -22,6 +22,11 @@
 			<description>
 			</description>
 		</method>
+		<method name="_get_class_icon_path" qualifiers="virtual const">
+			<return type="String" />
+			<description>
+			</description>
```
Comment (KoBeWi) [inline doc/classes/ScriptExtension.xml:28]: Needs description :v

**Comment by KoBeWi** [comment]
> The reason why I'm still skeptical here, is because this doesn't allow for multiple icons per one script (in case it contains multiple classes).

How would that even work? Only one icon can be displayed.

This looks interesting, but the method technically only exists to fetch a built-in script icon path. Adding it to Script class is overkill, because only GDScript can be built-in (and VisualScript, but it's irrelevant). I think you could bind this method only in GDScript and use `call()`.

**Comment by YuriSizov** [comment]
> How would that even work? Only one icon can be displayed.

For built-in scripts it's unlikely, but for discrete scripts it's not impossible that in the future we allow to add icons to sub-classes as well.

> the method technically only exists to fetch a built-in script icon path

So far, yeah. But there is nothing inherently limiting it to built-in scripts. So I preferred to make a general-purpose API. It also allows us to avoid binding EditorData and the GDScript module together. (Well, I haven't checked, maybe we already do, but that's not great if so.)

> I think you could bind this method only in GDScript and use `call()`.

Ah, yes, that could work. If the server-level change is rejected, this is a good fallback option.

**Comment by KoBeWi** [comment]
> but for discrete scripts it's not impossible that in the future we allow to add icons to sub-classes as well.

But that's still one icon per Create Dialog/Scene Tree/Inspector/Editor Help entry.
Built-in scripts also allow sub-classes. And sub-classes are technically separate scripts. So "multiple icons per script" does not make much sense.

> But there is nothing inherently limiting it to built-in scripts.

Well, use-cases are limited. We already have methods for fetching script icons. It would be more useful if you refactored the existing code to use `get_class_icon_path()`.

> it's not impossible that in the future
> So far, yeah

https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#to-solve-the-problem-it-has-to-exist-in-the-first-place
:v

**Comment by YuriSizov** [comment]
> Well, use-cases are limited. We already have methods for fetching script icons. It would be more useful if you refactored the existing code to use `get_class_icon_path()`.

Do you propose I refactor the entire EditorData caching system just to justify this API?

> But that's still one icon per Create Dialog/Scene Tree/Inspector/Editor Help entry.
> Built-in scripts also allow sub-classes. And sub-classes are technically separate scripts. So "multiple icons per script" does not make much sense.

The idea is that sub-classes will be registered independently and will have their own icons. GDScript does kind of treat them as separate scripts, but to the engine it's all one resource, and all under the same path in any cache we have.

> https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#to-solve-the-problem-it-has-to-exist-in-the-first-place
> :v

I personally don't read this as "Add hacks and then maybe refactor later" 🙃 Besides, I'm not adding any new functionality. We *do have* a ScriptLanguage method that allows us to retrieve a class icon already (`get_global_class_name`). It's just bundled together with a few other features into a single method that does 3 different things, and is only available to non-built-in scripts.

**Comment by YuriSizov** [comment]
PS. All this talk about subclasses is only relevant if we want to future-proof the API. But I'd be okay with it as is too. It's a PoC anyway. We can discard it altogether.

**Comment by anvilfolk** [comment]
> PS. All this talk about subclasses is only relevant if we want to future-proof the API. But I'd be okay with it as is too. It's a PoC anyway. We can discard it altogether.

For what it's worth, I'm going to try to make sure that inspector tooltips are working with subclasses also. Who knows if that'll work, but worth a try?

**Comment by dalexeev** [comment]
Please remove the outdated note in the [`@icon`](https://docs.godotengine.org/en/stable/classes/class_%40gdscript.html#class-gdscript-annotation-icon) docs (I'm just reminding you):

> The script must be registered as a global class using the class_name keyword for this to have a visible effect.

**Comment by YuriSizov** [comment]
Removed the note and adjusted the condition to handle any unnamed script, not just built-ins.

File: modules/gdscript/gdscript.cpp [hunk 2]
```diff
@@ -639,6 +639,26 @@ void GDScript::_update_doc() {
 
 	_add_doc(doc);
 }
+
+String GDScript::get_class_icon_path() const {
+	GDScriptParser parser;
+	GDScriptAnalyzer analyzer(&parser);
```
[the remark this answers is not in the document]
Comment (YuriSizov) [reply modules/gdscript/gdscript.cpp:645]: You're right, not sure where I got that idea from.

**Comment by anvilfolk** [comment]
> > PS. All this talk about subclasses is only relevant if we want to future-proof the API. But I'd be okay with it as is too. It's a PoC anyway. We can discard it altogether.
> 
> For what it's worth, I'm going to try to make sure that inspector tooltips are working with subclasses also. Who knows if that'll work, but worth a try?

I also just remembered that I looked into this but there was actually no way to get inner classes to show up in the inspector (unless you got into remote inspect mode when the app is running, but then no docstrings actually show up and everything has weird names), so that shouldn't be a consideration anymore either :)

File: modules/mono/csharp_script.h [hunk 3]
```diff
@@ -161,6 +161,9 @@ class CSharpScript : public Script {
 		Vector<DocData::ClassDoc> docs;
 		return docs;
 	}
+	virtual String get_class_icon_path() const override {
+		return String();
```
Comment (raulsntos) [inline modules/mono/csharp_script.h:165]: Now that https://github.com/godotengine/godot/pull/72619 has been merged.

```suggestion
		return icon_path;
```

File: doc/classes/ScriptExtension.xml (hunk 1 above)
Comment (KoBeWi) [inline doc/classes/ScriptExtension.xml:28]: Needs description.

**Comment by YuriSizov** [comment]
Rebased and tested again. Seems to work as before, and doesn't introduce any regression with mismatched types (fixed by https://github.com/godotengine/godot/pull/79203) for built-in scripts (the "Control" here has a script that extends Node):

image

File: doc/classes/ScriptExtension.xml (hunk 1 above)
Comment (KoBeWi) [reply doc/classes/ScriptExtension.xml:28]: I'm for enforcing writing descriptions for every added documentation field, regardless if the class is documented overall or not, and whether anyone can write it. Leaving empty fields like that is the reason why we have so much undocumented stuff.

> in the same consistent style

You mean the Godot documentation style used everywhere?

**Review by dalexeev — COMMENTED** [review]
I tested it a bit, it looks like it works fine.

Potentially this should work for inner classes as well? (However, you cannot attach an inner script in the editor, only create it via code.) I tried to do this, but it doesn't work. Not so important, just a note for possible future PR.

<details>
<summary>My try diff</summary>

```diff
diff --git a/modules/gdscript/gdscript_parser.cpp b/modules/gdscript/gdscript_parser.cpp
index debc85ebbf..e568592299 100644
--- a/modules/gdscript/gdscript_parser.cpp
+++ b/modules/gdscript/gdscript_parser.cpp
@@ -84,7 +84,7 @@ GDScriptParser::GDScriptParser() {
 	// Register valid annotations.
 	// TODO: Should this be static?
 	register_annotation(MethodInfo("@tool"), AnnotationInfo::SCRIPT, &GDScriptParser::tool_annotation);
-	register_annotation(MethodInfo("@icon", PropertyInfo(Variant::STRING, "icon_path")), AnnotationInfo::SCRIPT, &GDScriptParser::icon_annotation);
+	register_annotation(MethodInfo("@icon", PropertyInfo(Variant::STRING, "icon_path")), AnnotationInfo::SCRIPT | AnnotationInfo::CLASS, &GDScriptParser::icon_annotation);
 	register_annotation(MethodInfo("@static_unload"), AnnotationInfo::SCRIPT, &GDScriptParser::static_unload_annotation);
 
 	register_annotation(MethodInfo("@onready"), AnnotationInfo::VARIABLE, &GDScriptParser::onready_annotation);
@@ -762,6 +762,13 @@ void GDScriptParser::parse_class_member(T *(GDScriptParser::*p_parse_function)(b
 	}
 
 	for (AnnotationNode *&annotation : annotations) {
+		if constexpr (std::is_same_v<T, ClassNode>) {
+			if (annotation->name == SNAME("@icon")) {
+				// `@icon` needs to be applied in the parser.
+				annotation->apply(this, member);
+				continue;
+			}
+		}
 		member->annotations.push_back(annotation);
 	}
```

</details>

File: modules/gdscript/gdscript.cpp [hunk 4]
```diff
@@ -482,6 +482,25 @@ void GDScript::_clear_doc() {
 	docs.clear();
 	doc = DocData::ClassDoc();
 }
+
+String GDScript::get_class_icon_path() const {
+	GDScriptParser parser;
+	Error err = parser.parse(source, path, false);
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply modules/gdscript/gdscript.cpp:488]: The parent method is already cached.

[the remark this answers is not in the document]
Comment (YuriSizov) [reply modules/gdscript/gdscript.cpp:488]: `EditorData::get_script_icon` caches whatever it can resolve, so until the cache is invalidated. Though at this moment we invalidate it very aggressively, but it's a separate problem. Of course, if this can be cached on GDScript side too that'd be great, but I have no idea how to approach that.

[the remark this answers is not in the document]
Comment (dalexeev) [reply modules/gdscript/gdscript.cpp:488]: > Of course, if this can be cached on GDScript side too that'd be great, but I have no idea how to approach that.

Done: https://github.com/dalexeev/godot/commit/096560f31889a0048bbdb83fa834fabfb8144124.

**Review by raulsntos — COMMENTED** [review]
Tested with C# and it works as expected, but the icons don't update until the scene is reopened (same with GDScript).

File: doc/classes/ScriptExtension.xml (hunk 1 above)
Comment (KoBeWi) [reply doc/classes/ScriptExtension.xml:28]:
```suggestion
			Returns the path to the script's icon texture, or empty string if the script has no custom icon.
			</description>
```

Comment (akien-mga) [reply doc/classes/ScriptExtension.xml:28]: That's not what this method does though, it's what it should do in your implementation :)

Comment (dsnopek) [reply doc/classes/ScriptExtension.xml:28]: I'm not sure the `ScriptExtension` class should even be exposed? You should never interact with it from GDScript or C#, and even GDExtension dosen't directly interact with it (from GDExtension you register some callbacks via function in `gdextension_interface.h` which get used by `ScriptExtension`).

If we edit `register_engine_classes.cpp` and remove the line:

```c++
ClassDB::register_engine_class<ScriptExtension>();
```

... I think (although, I didn't test) that it should still work fine for it's purpose, except as an "unexposed" class.

And then I think (although, again I didn't test) that `godot --doctool` should try to generate docs for it.

Comment (dsnopek) [reply doc/classes/ScriptExtension.xml:28]: Er, sorry, I got myself mixed up! I was thinking of `ScriptInstanceExtension` which GDExtension developers don't directly interact with. For `ScriptExtension`, the GDExtension developer _does_ directly interact with the class, so this should remain exposed.

File: core/object/script_language_extension.h [hunk 5]
```diff
@@ -89,6 +90,12 @@ class ScriptExtension : public Script {
 
 		return class_doc;
 	}
+
+	virtual String get_class_icon_path() const override {
+		String ret;
+		GDVIRTUAL_REQUIRED_CALL(_get_class_icon_path, ret);
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply core/object/script_language_extension.h:96]: This method is called for every node with unique script, so at worst you will get a one-time error spam when opening a scene.

[the remark this answers is not in the document]
Comment (Bromeon) [reply core/object/script_language_extension.h:96]: > So long as Godot does something sane when empty `String` is returned, and it isn't getting called often enough that the error spam would be a problem, I think this would be OK?

That would essentially be an "implicit default implementation" (relying on certain behavior by not implementing the function) -- would it not be better to provide a default right away? Or do you see a big risk that people forget to implement this?



> This method is called for every node with unique script, so at worst you will get a one-time error spam when opening a scene.

Still, this is quite bad UX and there's nothing a developer of an older extension can do against it (except recompile extensions under newer Godot versions, which makes them incompatible with _older_ ones). Godot has generally a bit a problem with errors appearing on the console, to a user it's often unclear if it's their fault or just a bug in the engine.

[the remark this answers is not in the document]
Comment (YuriSizov) [reply core/object/script_language_extension.h:96]: Just to avoid unnecessary discussion, the only reason I made it a required call is because its cousin, `get_documentation` is. There is no deep reasoning for this and I can safely change it to a normal call. The implications weren't clear to me at the time.

I'll do the change once I'm back to work late next week.

File: modules/gdscript/gdscript.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (YuriSizov) [reply modules/gdscript/gdscript.cpp]: @dalexeev Included your patch and added you as a co-author!

**Comment by YuriSizov** [comment]
Added a GDScript-side cache per @dalexeev's suggestion and made the script extension method non-required for compatibility. At this point, unless there are further reviews, only the docs requirement is remaining. As I mentioned, I would prefer for it to be done in a separate PR, because no docs are present for the documented class yet. If it's a hard requirement, I'd prefer to do it myself as a prerequisite of this PR, otherwise I'd leave it for someone to cover in a follow-up.

**Review by dalexeev — APPROVED** [review]
Looks good to me. Since the GDScript changes are made by me, it would be nice to get a double check from @vnen or @adamscott, so that it doesn't feel like self-approve. But even if we find a bug later, we can always fix it, it doesn't look like a critical part that is worth the long hesitation. Let's give it a green light!

File: modules/gdscript/gdscript_parser.cpp [hunk 6]
```diff
@@ -3836,18 +3836,31 @@ bool GDScriptParser::tool_annotation(const AnnotationNode *p_annotation, Node *p
...
+	p_class->icon_path = path;
+
+	if (path.is_empty() || path.is_absolute_path()) {
+		p_class->simplified_icon_path = path.simplify_path();
+	} else if (path.is_relative_path()) {
+		p_class->simplified_icon_path = script_path.get_base_dir().path_join(path).simplify_path();
+	} else {
+		p_class->simplified_icon_path = path;
+	}
+
```
Comment (adamscott) [inline modules/gdscript/gdscript_parser.cpp:3863]: @dalexeev Shouldn't `simplified_icon_path` be instead a return value of a `get_simplified_icon_path()`, as the simplified path is necessarily derived from a path.

Comment (dalexeev) [reply modules/gdscript/gdscript_parser.cpp:3863]: Sorry, I didn't understand what you mean, could you rephrase it please?

Comment (adamscott) [reply modules/gdscript/gdscript_parser.cpp:3863]: I mean, a function that returns the simplified path
```cpp
String GDScriptParser::get_simplified_icon_path() {
	if (path.is_empty() || path.is_absolute_path()) {
		return path.simplify_path();
	} else if (path.is_relative_path()) {
		return script_path.get_base_dir().path_join(path).simplify_path();
	}
	return path;
}
```

Comment (YuriSizov) [reply modules/gdscript/gdscript_parser.cpp:3863]: This makes sense to me as well, resolving then :)

**Review by akien-mga — APPROVED** [review]
Looks good to me.

If there's no objection / requested changes, this should be mergeable for 4.2-dev5.

**Comment by akien-mga** [comment]
Thanks!
